File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Spots/play-entry-spot.ts
import { Injectable } from '@nestjs/common';
import { Inject } from '@nestjs/common';
import { ZLINK_ACTOR_GATEWAY } from '@zlink-systems/nestjs';
import { PlayActorJoinGameHandler } from './Handlers/play-actor-join-game-handler';
import { PacketNames, observeMilestoneRes, winMilestoneNotify } from '../../../../../Shared/Contracts/messages';
import { PlayerWinMilestoneEventHandler } from './Handlers/player-win-milestone-event-handler';
import { SampleNames } from '../../../../Configuration/sample-settings';
import type {
  ActorRef,
  ZLinkActorGateway,
  ZLinkActor,
  ZLinkEntrySpot,
  ZLinkEntrySpotContext,
  ZLinkMessage
} from '@zlink-systems/framework';
import type {
  JoinGameRes,
  PlayerInfo,
  PlayerWinMilestoneEvent,
  TicTacToeGameJoinReq,
  TicTacToeActor
} from '../../../../../Shared/Contracts/messages';

type PlayEntrySpotActor = TicTacToeActor & ZLinkActor;

@Injectable()
class MilestoneObserverRegistry {
  private readonly observers = new Map<string, PlayEntrySpotActor>();

  subscribe(actor: PlayEntrySpotActor): void {
    this.observers.set(actor.actorId, actor);
  }

  remove(actor: PlayEntrySpotActor): void {
    this.observers.delete(actor.actorId);
  }

  async notify(event: PlayerWinMilestoneEvent, receivingSpotNodeRid: string): Promise<void> {
    const payload = winMilestoneNotify(event, receivingSpotNodeRid);
    for (const actor of [...this.observers.values()]) {
      await actor.push(PacketNames.winMilestoneNotify, payload);
    }
  }
}

@Injectable()
class PlayEntrySpot implements ZLinkEntrySpot<PlayEntrySpotActor> {
  readonly context!: ZLinkEntrySpotContext<PlayEntrySpotActor>;

  constructor(
    private readonly milestoneObservers: MilestoneObserverRegistry,
    @Inject(ZLINK_ACTOR_GATEWAY) private readonly actorGateway: ZLinkActorGateway
  ) {}

  configure(): void {
    this.context.handlers.actorRequest(PacketNames.joinGameReq, PlayActorJoinGameHandler);
    this.context.handlers.addSubscribe(PlayerWinMilestoneEventHandler, SampleNames.playerMilestoneTopic);
  }

  async join(actorRef: ActorRef, player: PlayerInfo, roomId: string): Promise<JoinGameRes> {
    const request: TicTacToeGameJoinReq = {
      roomId,
      player: {
        actorId: player.actorId,
        displayName: player.displayName,
        level: player.level,
        wins: player.wins
      }
    };
    const joined = await this.actorGateway.joinSpot(actorRef, roomId, request).submit<Partial<JoinGameRes & { error: string }>>();
    const reply = joined.reply ?? {};
    if (joined.resultCode !== 0) {
      throw new Error(reply.error ?? `Room '${roomId}' rejected actor '${actorRef.actorId}'.`);
    }
    return reply as JoinGameRes;
  }

  async observeMilestone(actor: PlayEntrySpotActor): Promise<unknown> {
    this.milestoneObservers.subscribe(actor);
    return observeMilestoneRes(true);
  }

  async notifyMilestone(event: PlayerWinMilestoneEvent): Promise<void> {
    await this.milestoneObservers.notify(event, String(this.context.nodeRid));
  }

  async onDisconnectActor(actor: PlayEntrySpotActor, signal?: AbortSignal): Promise<void> {
    void signal;
    actor.markDisconnected();
    this.milestoneObservers.remove(actor);
  }

  async onCreateActor(actor: PlayEntrySpotActor, createRequest: ZLinkMessage, signal?: AbortSignal): Promise<void> {
    void signal;
    const player = createRequest.decode<Partial<TicTacToeActor>>(Object as never);
    if (typeof player.displayName === 'string') {
      actor.displayName = player.displayName;
    }
    if (typeof player.level === 'number') {
      actor.level = player.level;
    }
    if (typeof player.wins === 'number') {
      actor.wins = player.wins;
    }
  }

  async onJoinedActor(actor: PlayEntrySpotActor, signal?: AbortSignal): Promise<void> {
    if (actor.destroyAfterEntrySpotJoin) {
      await this.context.destroyActor(actor, signal);
    }
  }

  async onLeaveActor(actor: PlayEntrySpotActor, signal?: AbortSignal): Promise<void> {
    void signal;
    console.log(`entry spot: actor left. actor=${actor.actorId}`);
    this.milestoneObservers.remove(actor);
  }

}

Inject(MilestoneObserverRegistry)(PlayEntrySpot, undefined, 0);
Inject(ZLINK_ACTOR_GATEWAY)(PlayEntrySpot, undefined, 1);

export { MilestoneObserverRegistry, PlayEntrySpot };

File: framework/languages/dotnet/samples/TicTacToe/Server/Play/Infrastructure/ZLink/Spots/EntrySpot/PlayEntrySpot.cs
using TicTacToe.Server.Play.Infrastructure.ZLink.Actors;
using TicTacToe.Server.Play.Infrastructure.ZLink.Spots.EntrySpot.Handlers;
using TicTacToe.Server.Configuration;
using TicTacToe.Shared.Contracts;
using Zlink.Framework.Contracts.Messaging;
using Zlink.Framework.Contracts.Spots;

namespace TicTacToe.Server.Play.Infrastructure.ZLink.Spots.EntrySpot;

internal sealed class PlayEntrySpot(
    IZLinkEntrySpotContext context,
    ILogger<PlayEntrySpot> logger) : IZLinkEntrySpot<PlayActor>
{
    private readonly MilestoneObserverRegistry _milestoneObservers = new();

    public IZLinkEntrySpotContext Context { get; } = context;

    public void Configure()
    {
        Context.Handlers.AddActorPacket<PlayActorJoinGameHandler, PlayActor>();
        Context.Handlers.AddActorPacket<PlayActorObserveMilestoneHandler, PlayActor>();
        Context.Handlers.AddSubscribe<PlayerWinMilestoneEventHandler>(SampleTopics.PlayerMilestone);
    }

    public ValueTask OnCreateActorAsync(
        PlayActor actor,
        ZLinkMessage createRequest,
        CancellationToken cancellationToken)
    {
        actor.ApplyPlayer(createRequest.Decode<PlayerInfo>());
        logger.LogInformation(
            "entry spot: actor created. actor={ActorId}",
            actor.ActorId);
        return ValueTask.CompletedTask;
    }

    public ValueTask<ZLinkSpotActorJoinResult> OnActorJoinAsync(
        string actorId,
        ZLinkMessage request,
        CancellationToken cancellationToken)
    {
        return ValueTask.FromResult(ZLinkSpotActorJoinResult.Accept(request));
    }

    public async ValueTask OnJoinedActorAsync(
        PlayActor actor,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "entry spot: actor joined. actor={ActorId}",
            actor.ActorId);
        if (!actor.DestroyAfterEntrySpotJoin) return;

        logger.LogInformation(
            "entry spot: actor destroy requested. actor={ActorId}",
            actor.ActorId);
        await Context.DestroyActorAsync(actor, cancellationToken);
        logger.LogInformation(
            "entry spot: actor destroy completed. actor={ActorId}",
            actor.ActorId);
    }

    public ValueTask OnLeaveActorAsync(
        PlayActor actor,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "entry spot: actor left. actor={ActorId}",
            actor.ActorId);
        _milestoneObservers.Remove(actor);
        return ValueTask.CompletedTask;
    }

    public ValueTask OnDisconnectActorAsync(
        PlayActor actor,
        CancellationToken cancellationToken)
    {
        actor.MarkDisconnected();
        _milestoneObservers.Remove(actor);
        logger.LogInformation(
            "entry spot: actor disconnected. actor={ActorId}",
            actor.ActorId);
        return ValueTask.CompletedTask;
    }

    public ValueTask SubscribeMilestoneAsync(
        PlayActor actor,
        CancellationToken cancellationToken)
    {
        _milestoneObservers.Subscribe(actor);
        logger.LogInformation(
            "entry spot: milestone observer subscribed. actor={ActorId}, nodeRid={NodeRid}",
            actor.ActorId,
            Context.NodeRid);
        return ValueTask.CompletedTask;
    }

    public async ValueTask NotifyMilestoneAsync(
        PlayerWinMilestoneEvent milestone,
        CancellationToken cancellationToken)
    {
        await _milestoneObservers.NotifyAsync(
            milestone,
            Context.NodeRid.ToString(),
            cancellationToken);
    }

    private sealed class MilestoneObserverRegistry
    {
        private readonly Dictionary<string, PlayActor> _observers = new(StringComparer.Ordinal);

        public void Subscribe(PlayActor actor)
        {
            _observers[actor.ActorId] = actor;
        }

        public void Remove(PlayActor actor)
        {
            _observers.Remove(actor.ActorId);
        }

        public ValueTask NotifyAsync(
            PlayerWinMilestoneEvent milestone,
            string receivingSpotNodeRid,
            CancellationToken cancellationToken)
        {
            var notify = new WinMilestoneNotify(
                milestone.RoomId,
                milestone.ActorId,
                milestone.DisplayName,
                milestone.Wins,
                receivingSpotNodeRid);

            var observers = _observers.Values.ToArray();
            foreach (var observer in observers)
                observer.Context.BoundSession.Send(notify)
                    .Submit(cancellationToken);

            return ValueTask.CompletedTask;
        }
    }
}
